# Text Summarization: Extractive & Abstractive

**Our task** is to build a text summarizer to summarize just one document of a single topic. We don't have any labels (like human-made summaries) to compare against and calculate metrics (which would've been ROUGE in that case). It therefore makes the output rather subjective, and the only objective quantitative measure we can present is the compression ratio - how much of the text have been retained after summarization:

$$\text{compression\_rate}=\frac{len(\text{summary\_characters})}{len(\text{text\_characters})}$$

For the **data pre-processing**, we'll be using both `nltk` and `spaCy`, both of which do the same thing but both are used for practice.

**The baseline for the extractive summarization** is going to be `lead-3`, followed by TF-IDF and finally `TextRank`/`LexRank`. 

`lead-3` (essentially printing the first 3 sentences) is chosen because documents written with the point up front: news, reports, legal descriptions like this one, put the summary in the opening. That single passage contains the document's entire thesis. A frequency algorithm has to work hard to beat "just read the beginning."

TF-IDF, as the next simple but more complex logic, is comparable. We measure importance score per sentence based on the frequency of words appearing per sentence compared to the whole text, then sort and choose $30\%$ best of them. This method looks at a sentence in isolation and asks **"How many high-frequency words does this specific sentence have?"**.

`TextRank`/`LexRank` is the next level. They are based on Google's original PageRank algorithm, which ranked web pages based on how many other web pages linked to them. They look at the relationships between sentences and ask **"How much does this sentence look like all the other sentences in the document?"**. The algorithms create a square grid of size $N \times N$, where $N$ is the number of sentences. A sentence gets a high score if it is highly connected to other highly connected sentences. It represents the "consensus" of the text. 

The only difference is how each algorithm calculates the scores. `TextRank` uses the word overlap formula (how many common words two sentences share, normalized by their length), then the PageRank algorithm takes this static map (the matrix) and simulates a "random surfer" clicking from sentence to sentence based on those similarity weights. It multiplies the matrix by a vector over and over again until the scores stabilize and it figures out which sentences are the "hubs" that all the other sentences revolve around.

`TextRank`'s flaw is that it treats all words equally (sharing the word "orbiter" counts exactly the same as sharing the word "said"), `LexRank` fixes this by using TF-IDF and Cosine Similarity. Instead of just counting overlapping words, `LexRank` converts every sentence into a TF-IDF vector, it then calculates the exact geometric angle (Cosine Similarity) between those sentence vectors to build the $N \times N$ matrix.

For the **abstractive summarization**, we're going to use the middle ground. The (subjectively) best and concise output of the extractive summarization will be used as input for a transformer.

## Data preparation

The task calls for using both `SpaCy` and `nltk`, both of which do the same job — split into sentences, split into words. There is no reason to use both. We need to choose the more appropriate option based on the data, which, as mentioned, is a single document and a single topic.


| **Feature** | **NLTK** (2001, academic) | **spaCy** (2015, industrial) |
| :--- | :--- | :--- |
| **Design Philosophy**<br>(purpose) | A toolbox. We have to call separate functions for every single step. | A pipeline. We feed text in once, and it returns an object with everything calculated. |
| **Concept** | An unsupervised statistical model (`punkt`). It counts frequencies and looks for patterns in text. It builds a list of abbreviations (like _Mr._, _Dr._, or _Inc._) by calculating how likely a word is to appear before a period without ending a sentence.<br><br>Very fast, lightweight, requires very little computer memory, but doesn't understand the context or grammar. | A Deep Neural Network (CNN/transformer-based depending on the choice) parser that analyzes the grammatical structure of the entire text simultaneously. It builds a structural tree of the whole sentence. It figures out which words are subjects, verbs, and objects, and how they connect to each other; identifies sentence splits as a byproduct of understanding the overall syntax.<br><br>Highly accurate on complex, messy, or conversational text because it understands the full sentence context, not just punctuation rules, but computationally heavier, slower and requires more memory. |
| **Part of Speech (POS) Filtering** | Manual step. We must explicitly call `nltk.pos_tag()`. | Automatic. The `pos_` property (like `PROPN` or `VERB`) is ready instantly. |
| **Lemmatization**<br>(e.g., fly, flew, flown treated as a single entity) | Weak by default. Requires us to manually run a POS tagger first. | Strong and automatic. The `lemma_` property is immediately available for every word. |
| **Named Entities (NER)** | Clunky. Requires chaining several different tokenizers and chunkers. | Automatic. Entities like "NASA" or "1984" are immediately tagged in `doc.ents`. |
| **Best Use Case** | Academic learning or string manipulation. | Industrial applications, extracting deep meaning, and building actual products. |


Being simpler, lighter and faster, `nltk` might become our "data pre-processing baseline", but the concept of a baseline usually refers to something standard and param-default rather than a manually configurable drag. Even at this stage, it gives us no practical benefit. Unless we have to use heavy custom logic, I wouldn't go back to this option. We implement it only because I think it's a good show-case for learning.

### Data & constants

In [1]:
TEXT = "The Orbiter Discovery, OV-103, is considered eligible for listing in the National Register of Historic Places (NRHP) in the context of the U.S. Space Shuttle Program (1969-2011) under Criterion A in the areas of Space Exploration and Transportation and under Criterion C in the area of Engineering. Because it has achieved significance within the past fifty years, Criteria Consideration G applies. Under Criterion A, Discovery is significant as the oldest of the three extant orbiter vehicles constructed for the Space Shuttle Program (SSP), the longest running American space program to date; she was the third of five orbiters built by NASA. Unlike the Mercury, Gemini, and Apollo programs, the SSP’s emphasis was on cost effectiveness and reusability, and eventually the construction of a space station. Including her maiden voyage (launched August 30, 1984), Discovery flew to space thirty-nine times, more than any of the other four orbiters; she was also the first orbiter to fly twenty missions. She had the honor of being chosen as the Return to Flight vehicle after both the Challenger and Columbia accidents. Discovery was the first shuttle to fly with the redesigned SRBs, a result of the Challenger accident, and the first shuttle to fly with the Phase II and Block I SSME. Discovery also carried the Hubble Space Telescope to orbit and performed two of the five servicing missions to the observatory. She flew the first and last dedicated Department of Defense (DoD) missions, as well as the first unclassified defense-related mission. In addition, Discovery was vital to the construction of the International Space Station (ISS); she flew thirteen of the thirty-seven total missions flown to the station by a U.S. Space Shuttle. She was the first orbiter to dock to the ISS, and the first to perform an exchange of a resident crew. Under Criterion C, Discovery is significant as a feat of engineering. According to Wayne Hale, a flight director from Johnson Space Center, the Space Shuttle orbiter represents a “huge technological leap from expendable rockets and capsules to a reusable, winged, hypersonic, cargo-carrying spacecraft.” Although her base structure followed a conventional aircraft design, she used advanced materials that both minimized her weight for cargo-carrying purposes and featured low thermal expansion ratios, which provided a stable base for her Thermal Protection System (TPS) materials. The Space Shuttle orbiter also featured the first reusable TPS; all previous spaceflight vehicles had a single-use, ablative heat shield. Other notable engineering achievements of the orbiter included the first reusable orbital propulsion system, and the first two-fault-tolerant Integrated Avionics System. As Hale stated, the Space Shuttle remains “the largest, fastest, winged hypersonic aircraft in history,” having regularly flown at twenty-five times the speed of sound."  # noqa: RUF001
SUMMARIZATION_COEF = 0.3

In order to process text, we need to encode, and to encode it needs to be tokenized. The first question to ask is *what should one token be?*. We can tokenize into characters, words, morphemes. But **extractive summarization** itself is purely statistical. It counts frequencies and similarities giving the exact verbatim pieces, we have to use words to properly measure the similarities and count occurrences. That's why **extractive summarization** uses words as tokens.

Then, the next thing to address is **the language**. The text is in English, we want to divide it into words, but first into sentences, this is called **sentence boundary detection (splitting)**. We can't just split it by periods (`.`) because there are proper nouns like `U.S.` or rely on capital layers for that reason; there are also abbreviations. `nltk` and `SpaCy` have pre-trained unsupervised ML models to split properly.

To consider how to handle **Proper Nouns and Named Entities** (like *Discovery* or *NASA*), as they carry the highest informational weight in a text, we will utilize **Part of Speech (POS) filtering**. By leveraging `nltk` and `SpaCy`'s grammar tagging, we can explicitly retain only Proper Nouns, Common Nouns, Verbs, and Adjectives. Additionally, `SpaCy`'s out-of-the-box **Named Entity Recognition (NER)** allows us to identify organizations, dates, and locations, which can optionally be used to give higher importance scores to sentences containing concrete facts.

There are also tenses and irregular verbs in the English language that are all supposed to be treated equally, we need lemmatization (supplied by `nltk` and `SpaCy`), it will treat `flying`, `flew`, `flown`, `fly`" as just `fly` correctly capturing the context. **Case normalization** (lowercasing) also contributes to the solution of this problem.

Typical high-frequency words called stop-words like `the`, `a`, `an`, `will`, `re`, `across`, `not` etc. have to be stripped as they don't express what the text is about not to appear in the top useful. `nltk` and `SpaCy` supply the ready-made lists.

Punctuation also doesn't help in **extractive summarization** and shouldn't become tokens, we will be removing them, the libraries also provide punctuation sets.

After splitting into sentences, "fitting" the lead-3 baseline becomes viable.

**For TF-IDF** strategy:

$$\text{tf-idf}(t, d) = \underbrace{\text{tf}(t,d)}_{\substack{\text{how often term } t \\ \text{appears in \emph{this} doc}}} \times \underbrace{\text{idf}(t)}_{\substack{\text{how \emph{rare} } t \text{ is} \\ \text{across all docs}}}$$

$$\text{tf}(t,d)=\frac{\text{Number of times term }t\text{ appears in document }d}{\text{Total number of terms in document }d}$$

$$\begin{gather*}
\text{idf}(t) = \ln{\left(\frac{1+N}{1+\text{df}(t)}\right)} + 1 \\ \\ 
	\begin{aligned} 
		&\bullet \text{ $N$ — total number of documents} \\
		&\bullet \text{ $\text{df}(t)$ — \emph{document frequency}: how many documents contain $t$ at all} \\
		&\bullet \text{ the $+1$s — smoothing, so a term in every document doesn't give $\ln(1)=0$ and vanish entirely} 
	\end{aligned} 
\end{gather*}$$

We only have the **TF** part because there is only one document. Usually, **TF-IDF** is used on documents and rank how well they fit certain query (search engines). And the reason the **IDF** part is there is because it measures how rare a term is across all the documents. Therefore, we just normalize be sentence length:

$$\begin{gather*}
\text{score}(\text{sentence}) = \frac{1}{\text{sent\_len}}\sum_{w \in \text{sent}} \text{freq}(w), \qquad \text{freq}(w) = \frac{\text{count}(w)}{\text{total\_count}(v)} \\ \\
\begin{aligned}
&\bullet \text{ count($w$) - word count within a sentence}  \\
&\bullet \text{ total\_count($v$) - word count across the whole vocabulary set (document)}  \\
\end{aligned}
\end{gather*}$$

## Pre-processing

### nltk

In [2]:
import string

import nltk
from nltk.corpus import stopwords, wordnet
from nltk.stem import WordNetLemmatizer
from nltk.tokenize import sent_tokenize, word_tokenize

In [3]:
# NLTK ships without data - download once
"""
• stopwords - the list of uninformative words like articles, particles etc.
• wordnet - to convert NER labels
• punkt_tab, averaged_perceptron_tagger_eng - for sent_tokenize to work
"""
for pkg in ['stopwords', 'wordnet', 'punkt_tab', 'averaged_perceptron_tagger_eng']:
    nltk.download(pkg, quiet=True)

In [4]:
print(f"10 stop words examples: {stopwords.words('english')[:10]}")
print(f"Punctuation examples: {string.punctuation}")

10 stop words examples: ['a', 'about', 'above', 'after', 'again', 'against', 'ain', 'all', 'am', 'an']
Punctuation examples: !"#$%&'()*+,-./:;<=>?@[\]^_`{|}~


In [5]:
nltk_stop_words = set(stopwords.words("english"))
punctuation = set(string.punctuation) | {"''", "``", "--", "...", "'s", "’", "“", "”"}  # noqa: RUF001
lemmatizer = WordNetLemmatizer()

# split into sentences using the nltk's pre-trained unsupervised sentence tokenizer
nltk_raw_sentences = sent_tokenize(TEXT)
nltk_raw_sentences

['The Orbiter Discovery, OV-103, is considered eligible for listing in the National Register of Historic Places (NRHP) in the context of the U.S. Space Shuttle Program (1969-2011) under Criterion A in the areas of Space Exploration and Transportation and under Criterion C in the area of Engineering.',
 'Because it has achieved significance within the past fifty years, Criteria Consideration G applies.',
 'Under Criterion A, Discovery is significant as the oldest of the three extant orbiter vehicles constructed for the Space Shuttle Program (SSP), the longest running American space program to date; she was the third of five orbiters built by NASA.',
 'Unlike the Mercury, Gemini, and Apollo programs, the SSP’s emphasis was on cost effectiveness and reusability, and eventually the construction of a space station.',
 'Including her maiden voyage (launched August 30, 1984), Discovery flew to space thirty-nine times, more than any of the other four orbiters; she was also the first orbiter to

The problem is that by default `nltk`'s lemmatizer assumes each word is a noun so we need to run POS (part of speech) tagger on the original data to get the proper part of speech. However, because `nltk` is an old package of algorithms evolved over decades, not all of its methods agree directly. We can't just get the part of speech from `nltk`'s POS and pass to the lemmatizer because the `pos_tag` was trained on the Penn Treebank dataset and lemmatizer - on `wordnet`. We need to map the targets:

In [6]:
def nltk_to_wordnet_pos(nltk_tag: str):
    # NOTE: pos_tag has way more parts of speech, but we don't need such detailed distinction
    return {"J": wordnet.ADJ, "V": wordnet.VERB,
            "N": wordnet.NOUN, "R": wordnet.ADV}.get(nltk_tag[0], wordnet.NOUN)

# NOTE: I might've needed to exclude ADV, but that's arguable

In [7]:
def nltk_tokenize(text):
    # split into unique words considering lemmatization, stop words and NER filtering
    words = []
    word_candidates = word_tokenize(text) # includes raw words, prepositions, punctuation etc.
    tagged_words = nltk.pos_tag(word_candidates) # this returns pairs like: [('Discovery', 'NNP'), ('flew', 'VBD')]
    for raw_w, nltk_tag in tagged_words:
        raw_w_lower = raw_w.lower()

        # if a word is a stop word or punctuation, skip it
        if (raw_w_lower in nltk_stop_words) or (raw_w_lower in punctuation):
            continue

        words.append(lemmatizer.lemmatize(raw_w_lower, nltk_to_wordnet_pos(nltk_tag)))
    return words

nltk_words = nltk_tokenize(TEXT)
print(f"Tokens: {nltk_words}")

Tokens: ['orbiter', 'discovery', 'ov-103', 'consider', 'eligible', 'list', 'national', 'register', 'historic', 'place', 'nrhp', 'context', 'u.s.', 'space', 'shuttle', 'program', '1969-2011', 'criterion', 'area', 'space', 'exploration', 'transportation', 'criterion', 'c', 'area', 'engineering', 'achieve', 'significance', 'within', 'past', 'fifty', 'year', 'criterion', 'consideration', 'g', 'applies', 'criterion', 'discovery', 'significant', 'old', 'three', 'extant', 'orbiter', 'vehicle', 'construct', 'space', 'shuttle', 'program', 'ssp', 'long', 'running', 'american', 'space', 'program', 'date', 'third', 'five', 'orbiter', 'build', 'nasa', 'unlike', 'mercury', 'gemini', 'apollo', 'program', 'ssp', 'emphasis', 'cost', 'effectiveness', 'reusability', 'eventually', 'construction', 'space', 'station', 'include', 'maiden', 'voyage', 'launch', 'august', '30', '1984', 'discovery', 'fly', 'space', 'thirty-nine', 'time', 'four', 'orbiter', 'also', 'first', 'orbiter', 'fly', 'twenty', 'mission', 

Successfully tokenized.

### SpaCy

In [8]:
import spacy

Unlike `nltk`, `SpaCy` is an automated pipeline that features different deep NN models. The first step is to choose the appropriate model for the task.

|         | size    | based on               | word vectors (embeddings)                       | NER F     | sentence F (splitting sentences)|
| ------- | ------- | ---------------------- | ---------------------------------- | --------- | ---------- |
| **sm**  | 14.5 MB | CNN (tok2vec)          | **none**                           | 0.843     | **0.907**  |
| **md**  | ~40 MB  | CNN, same architecture | 20,000 unique 300-dimension vectors | 0.843     | 0.907      |
| **lg**| ~560 MB | CNN, same architecture | ~500,000 unique 300-dimension vectors             | 0.844     | 0.907      |
| **trf** | ~440 MB | **RoBERTa-base**       | none                               | **0.901** | 0.897      |

`sm`, `md` and `lg` have essentially identical accuracy. `SpaCy`'s own maintainers published accuracy tables for the three are the same except NER recall. They do not differ by quality. They differ by **whether they ship word vectors**:
- Need `token.vector`, `doc.similarity()`, semantic comparison → **md or lg**
- Need only tokenization, POS, lemma, NER, sentence splits → **sm for identical result + cheaper**

`trf` (transformer-based, most accurate, slowest) is the only real accuracy jump, and only for NER (0.843 → 0.901) and parsing.

In [9]:
# !python -m spacy download en_core_web_sm

In [10]:
# the splitting model
nlp = spacy.load("en_core_web_sm")

In [11]:
# split into sentences using the spacy's pre-trained NN
spacy_raw_text = nlp(TEXT)
spacy_raw_sentences = list(spacy_raw_text.sents)      # all document sentences, IN DOCUMENT ORDER
spacy_raw_sentences

[The Orbiter Discovery, OV-103, is considered eligible for listing in the National Register of Historic Places (NRHP) in the context of the U.S. Space Shuttle Program (1969-2011) under Criterion A in the areas of Space Exploration and Transportation and under Criterion C in the area of Engineering.,
 Because it has achieved significance within the past fifty years, Criteria Consideration G applies.,
 Under Criterion A, Discovery is significant as the oldest of the three extant orbiter vehicles constructed for the Space Shuttle Program (SSP), the longest running American space program to date; she was the third of five orbiters built by NASA.,
 Unlike the Mercury, Gemini, and Apollo programs, the SSP’s emphasis was on cost effectiveness and reusability, and eventually the construction of a space station.,
 Including her maiden voyage (launched August 30, 1984), Discovery flew to space thirty-nine times, more than any of the other four orbiters; she was also the first orbiter to fly twen

In [12]:
# each sentence in spacy is "spacy.tokens.span.Span" object and in nltk they are just str
spacy_raw_sentences_str = [str(sent) for sent in spacy_raw_sentences]
set(spacy_raw_sentences_str) == set(nltk_raw_sentences)

True

The split is identical

In [62]:
# spacy "sm" model's stop words
spacy_stop_words = nlp.Defaults.stop_words

print(f"Stop words examples: {spacy_stop_words}")


from spacy.lang.punctuation import LIST_PUNCT

print(f"Punctuation examples: {LIST_PUNCT}")

Stop words examples: {'too', 'hereafter', 'under', '’d', 'wherein', 'five', 'amount', 'please', 'whom', 'afterwards', 'through', 'yourself', 'often', 'up', 'otherwise', 'part', 'everywhere', 'give', 'when', 'several', 'hence', 'move', 'herself', 'though', 'can', 'behind', 'further', '‘s', 'seem', 'so', 'seeming', 'third', 'your', 'beyond', 'thereby', 'either', 'top', 'from', 'forty', 'such', 'did', 'for', 'fifty', 'serious', 'anyway', 'above', 'eleven', 'latterly', 'perhaps', 'upon', 'why', 'his', 'along', 'latter', 'does', "'ll", 'or', 'a', 'whose', 'an', 'not', 'nothing', 'anything', 'could', 'the', 'per', 'to', 'her', 'whenever', 'ourselves', 'of', 'mine', 'regarding', 'almost', 'hers', 'ca', 'which', 'last', 'full', 'empty', 'fifteen', 'two', 'our', 'nine', 'themselves', 'their', 'four', "'ve", 'get', 'thus', 'used', 'by', 'in', 'whither', 'again', 'whence', 'go', 'no', "n't", 'next', 'although', 'around', 'yours', 'into', 'whoever', 'during', 'really', 'hereby', 'noone', 'n’t', 'e

In [14]:
def spacy_tokenize(text):
    words = []
    for token in nlp(str(text)):
        if token.is_stop or token.is_punct:
            continue
        words.append(token.lemma_.lower())
    return words

spacy_words = spacy_tokenize(TEXT)
print(f"Tokens: {spacy_words}")

Tokens: ['orbiter', 'discovery', 'ov-103', 'consider', 'eligible', 'list', 'national', 'register', 'historic', 'places', 'nrhp', 'context', 'u.s.', 'space', 'shuttle', 'program', '1969', '2011', 'criterion', 'area', 'space', 'exploration', 'transportation', 'criterion', 'c', 'area', 'engineering', 'achieve', 'significance', 'past', 'year', 'criteria', 'consideration', 'g', 'apply', 'criterion', 'discovery', 'significant', 'old', 'extant', 'orbiter', 'vehicle', 'construct', 'space', 'shuttle', 'program', 'ssp', 'long', 'running', 'american', 'space', 'program', 'date', 'orbiter', 'build', 'nasa', 'unlike', 'mercury', 'gemini', 'apollo', 'program', 'ssp', 'emphasis', 'cost', 'effectiveness', 'reusability', 'eventually', 'construction', 'space', 'station', 'include', 'maiden', 'voyage', 'launch', 'august', '30', '1984', 'discovery', 'fly', 'space', 'thirty', 'time', 'orbiter', 'orbiter', 'fly', 'mission', 'honor', 'choose', 'return', 'flight', 'vehicle', 'challenger', 'columbia', 'acciden

In [15]:
print(nltk_words)

['orbiter', 'discovery', 'ov-103', 'consider', 'eligible', 'list', 'national', 'register', 'historic', 'place', 'nrhp', 'context', 'u.s.', 'space', 'shuttle', 'program', '1969-2011', 'criterion', 'area', 'space', 'exploration', 'transportation', 'criterion', 'c', 'area', 'engineering', 'achieve', 'significance', 'within', 'past', 'fifty', 'year', 'criterion', 'consideration', 'g', 'applies', 'criterion', 'discovery', 'significant', 'old', 'three', 'extant', 'orbiter', 'vehicle', 'construct', 'space', 'shuttle', 'program', 'ssp', 'long', 'running', 'american', 'space', 'program', 'date', 'third', 'five', 'orbiter', 'build', 'nasa', 'unlike', 'mercury', 'gemini', 'apollo', 'program', 'ssp', 'emphasis', 'cost', 'effectiveness', 'reusability', 'eventually', 'construction', 'space', 'station', 'include', 'maiden', 'voyage', 'launch', 'august', '30', '1984', 'discovery', 'fly', 'space', 'thirty-nine', 'time', 'four', 'orbiter', 'also', 'first', 'orbiter', 'fly', 'twenty', 'mission', 'honor',

In [16]:
print(spacy_words)

['orbiter', 'discovery', 'ov-103', 'consider', 'eligible', 'list', 'national', 'register', 'historic', 'places', 'nrhp', 'context', 'u.s.', 'space', 'shuttle', 'program', '1969', '2011', 'criterion', 'area', 'space', 'exploration', 'transportation', 'criterion', 'c', 'area', 'engineering', 'achieve', 'significance', 'past', 'year', 'criteria', 'consideration', 'g', 'apply', 'criterion', 'discovery', 'significant', 'old', 'extant', 'orbiter', 'vehicle', 'construct', 'space', 'shuttle', 'program', 'ssp', 'long', 'running', 'american', 'space', 'program', 'date', 'orbiter', 'build', 'nasa', 'unlike', 'mercury', 'gemini', 'apollo', 'program', 'ssp', 'emphasis', 'cost', 'effectiveness', 'reusability', 'eventually', 'construction', 'space', 'station', 'include', 'maiden', 'voyage', 'launch', 'august', '30', '1984', 'discovery', 'fly', 'space', 'thirty', 'time', 'orbiter', 'orbiter', 'fly', 'mission', 'honor', 'choose', 'return', 'flight', 'vehicle', 'challenger', 'columbia', 'accident', 'dis

All the comparisons on the difference between the tokens are in the "Conclusion" section.

## Extractive Summarization

### Baseline

In [17]:
def lead_3_baseline(raw_sentences):
    return " ".join(raw_sentences[:3])

In [18]:
def compression_rate(original_text, summarized_text):
    return len(summarized_text) / len(original_text)

#### `nltk`'s lead-3

In [19]:
nltk_lead_3_summary = lead_3_baseline(nltk_raw_sentences)
nltk_lead_3_summary

'The Orbiter Discovery, OV-103, is considered eligible for listing in the National Register of Historic Places (NRHP) in the context of the U.S. Space Shuttle Program (1969-2011) under Criterion A in the areas of Space Exploration and Transportation and under Criterion C in the area of Engineering. Because it has achieved significance within the past fifty years, Criteria Consideration G applies. Under Criterion A, Discovery is significant as the oldest of the three extant orbiter vehicles constructed for the Space Shuttle Program (SSP), the longest running American space program to date; she was the third of five orbiters built by NASA.'

In [20]:
print(f"Compression rate: {compression_rate(TEXT, nltk_lead_3_summary)}")

Compression rate: 0.22161046111493463


#### `SpaCy`'s lead-3

In [21]:
spacy_lead_3_summary = lead_3_baseline(spacy_raw_sentences_str)
spacy_lead_3_summary

'The Orbiter Discovery, OV-103, is considered eligible for listing in the National Register of Historic Places (NRHP) in the context of the U.S. Space Shuttle Program (1969-2011) under Criterion A in the areas of Space Exploration and Transportation and under Criterion C in the area of Engineering. Because it has achieved significance within the past fifty years, Criteria Consideration G applies. Under Criterion A, Discovery is significant as the oldest of the three extant orbiter vehicles constructed for the Space Shuttle Program (SSP), the longest running American space program to date; she was the third of five orbiters built by NASA.'

In [22]:
print(f"Compression rate: {compression_rate(TEXT, spacy_lead_3_summary)}")

Compression rate: 0.22161046111493463


### TF (from TF-IDF)

#### nltk

In [23]:
def tf_frequencies(document_tokens):
    frequencies = {}
    for word in document_tokens:
        # Get current count (default to 0 if not found) and add 1
        frequencies[word] = frequencies.get(word, 0) + 1

    # normalization by total words
    frequencies = {k:v/len(document_tokens) for k,v in frequencies.items()}
    return frequencies

nltk_frequencies = tf_frequencies(nltk_words)
sorted(nltk_frequencies.items(), key=lambda item: item[1], reverse=True)

[('space', 0.04744525547445255),
 ('first', 0.0364963503649635),
 ('orbiter', 0.032846715328467155),
 ('shuttle', 0.029197080291970802),
 ('fly', 0.029197080291970802),
 ('discovery', 0.025547445255474453),
 ('criterion', 0.01824817518248175),
 ('mission', 0.01824817518248175),
 ('program', 0.014598540145985401),
 ('engineering', 0.010948905109489052),
 ('vehicle', 0.010948905109489052),
 ('station', 0.010948905109489052),
 ('also', 0.010948905109489052),
 ('reusable', 0.010948905109489052),
 ('system', 0.010948905109489052),
 ('u.s.', 0.0072992700729927005),
 ('area', 0.0072992700729927005),
 ('c', 0.0072992700729927005),
 ('significant', 0.0072992700729927005),
 ('ssp', 0.0072992700729927005),
 ('five', 0.0072992700729927005),
 ('construction', 0.0072992700729927005),
 ('include', 0.0072992700729927005),
 ('time', 0.0072992700729927005),
 ('flight', 0.0072992700729927005),
 ('challenger', 0.0072992700729927005),
 ('accident', 0.0072992700729927005),
 ('perform', 0.0072992700729927005

In [24]:
def tf_sentence_scores(raw_sentences, tokenizer, tf_frequencies):
    sentence_scores = {}

    for i, sent in enumerate(raw_sentences):
        sentence_scores[i] = 0
        sent_tokens = tokenizer(str(sent))

        # If raw_sentences contains a short sentence made entirely of stop-words or punctuation (e.g., "And so on."),
        if not sent_tokens:
            continue

        sentence_tokens_counts = {}
        for token in sent_tokens:
            sentence_tokens_counts[token] = sentence_tokens_counts.get(token, 0) + 1

        for word, count in sentence_tokens_counts.items():
            sentence_scores[i] += count * tf_frequencies.get(word, 0)

        sentence_scores[i] = sentence_scores[i] / len(sent_tokens)

    return sentence_scores

In [25]:
nltk_sentence_scores = tf_sentence_scores(nltk_raw_sentences, nltk_tokenize, nltk_frequencies)
nltk_sentence_scores_sorted = sorted(nltk_sentence_scores.items(), key=lambda item: item[1], reverse=True)
nltk_sentence_scores_sorted

[(9, 0.016220600162206),
 (6, 0.015967153284671527),
 (4, 0.015510948905109487),
 (10, 0.015004055150040552),
 (14, 0.014355231143552305),
 (2, 0.013990267639902675),
 (8, 0.012914093206064006),
 (0, 0.012352610892756875),
 (11, 0.0121654501216545),
 (15, 0.012165450121654499),
 (7, 0.01094890510948905),
 (12, 0.010948905109489048),
 (16, 0.0103048518677544),
 (3, 0.008602711157455683),
 (5, 0.005930656934306569),
 (13, 0.00527169505271695),
 (1, 0.0051094890510948905)]

In [26]:
def compose_summary_tf(raw_sentences: list[str], scored_sentences: list[tuple[int, float]], summary_coef: float = 0.3):
    # --- Select a fraction (30%) or at least 1 sentence ---
    select_length = max(1, int(len(raw_sentences) * summary_coef))

    top_30p_sentences = scored_sentences[:select_length]
    top_30p_indices = [i for i, sc in top_30p_sentences]

    # --- Restore DOCUMENT order ---
    # Because our items are just integer indices, restoring chronological order just means sorting
    # the numbers from lowest to highest.
    top_indices_sorted = sorted(top_30p_indices)

    # --- Build the summary ---
    return " ".join(str(raw_sentences[i]).strip() for i in top_indices_sorted)

In [29]:
nltk_tf_summary = compose_summary_tf(nltk_raw_sentences, nltk_sentence_scores_sorted, SUMMARIZATION_COEF)
nltk_tf_summary

'Including her maiden voyage (launched August 30, 1984), Discovery flew to space thirty-nine times, more than any of the other four orbiters; she was also the first orbiter to fly twenty missions. Discovery was the first shuttle to fly with the redesigned SRBs, a result of the Challenger accident, and the first shuttle to fly with the Phase II and Block I SSME. In addition, Discovery was vital to the construction of the International Space Station (ISS); she flew thirteen of the thirty-seven total missions flown to the station by a U.S. Space Shuttle. She was the first orbiter to dock to the ISS, and the first to perform an exchange of a resident crew. The Space Shuttle orbiter also featured the first reusable TPS; all previous spaceflight vehicles had a single-use, ablative heat shield.'

In [30]:
print(f"Compression rate: {compression_rate(TEXT, nltk_tf_summary)}")

Compression rate: 0.27426015141087406


#### SpaCy

In [31]:
spacy_frequencies = tf_frequencies(spacy_words)
sorted(spacy_frequencies.items(), key=lambda item: item[1], reverse=True)

[('space', 0.050980392156862744),
 ('orbiter', 0.03529411764705882),
 ('shuttle', 0.03137254901960784),
 ('fly', 0.03137254901960784),
 ('discovery', 0.027450980392156862),
 ('mission', 0.0196078431372549),
 ('program', 0.01568627450980392),
 ('criterion', 0.01568627450980392),
 ('engineering', 0.011764705882352941),
 ('vehicle', 0.011764705882352941),
 ('station', 0.011764705882352941),
 ('carry', 0.011764705882352941),
 ('reusable', 0.011764705882352941),
 ('system', 0.011764705882352941),
 ('u.s.', 0.00784313725490196),
 ('area', 0.00784313725490196),
 ('c', 0.00784313725490196),
 ('significant', 0.00784313725490196),
 ('ssp', 0.00784313725490196),
 ('construction', 0.00784313725490196),
 ('include', 0.00784313725490196),
 ('thirty', 0.00784313725490196),
 ('time', 0.00784313725490196),
 ('flight', 0.00784313725490196),
 ('challenger', 0.00784313725490196),
 ('accident', 0.00784313725490196),
 ('perform', 0.00784313725490196),
 ('defense', 0.00784313725490196),
 ('iss', 0.0078431372

In [32]:
spacy_sentence_scores = tf_sentence_scores(spacy_raw_sentences, spacy_tokenize, spacy_frequencies)
spacy_sentence_scores_sorted = sorted(spacy_sentence_scores.items(), key=lambda item: item[1], reverse=True)
spacy_sentence_scores_sorted

[(4, 0.017401960784313723),
 (9, 0.016924664602683177),
 (2, 0.01624649859943977),
 (6, 0.014005602240896352),
 (7, 0.013725490196078433),
 (14, 0.013165266106442574),
 (0, 0.012636165577342047),
 (11, 0.01241830065359477),
 (12, 0.011764705882352934),
 (16, 0.011072664359861588),
 (8, 0.010588235294117647),
 (10, 0.009523809523809523),
 (3, 0.009243697478991595),
 (15, 0.008683473389355743),
 (5, 0.006372549019607843),
 (13, 0.006033182503770738),
 (1, 0.00392156862745098)]

In [33]:
spacy_tf_summary = compose_summary_tf(spacy_raw_sentences, spacy_sentence_scores_sorted, SUMMARIZATION_COEF)
spacy_tf_summary

'Under Criterion A, Discovery is significant as the oldest of the three extant orbiter vehicles constructed for the Space Shuttle Program (SSP), the longest running American space program to date; she was the third of five orbiters built by NASA. Including her maiden voyage (launched August 30, 1984), Discovery flew to space thirty-nine times, more than any of the other four orbiters; she was also the first orbiter to fly twenty missions. Discovery was the first shuttle to fly with the redesigned SRBs, a result of the Challenger accident, and the first shuttle to fly with the Phase II and Block I SSME. Discovery also carried the Hubble Space Telescope to orbit and performed two of the five servicing missions to the observatory. In addition, Discovery was vital to the construction of the International Space Station (ISS); she flew thirteen of the thirty-seven total missions flown to the station by a U.S. Space Shuttle.'

In [34]:
print(f"Compression rate: {compression_rate(TEXT, spacy_tf_summary)}")

Compression rate: 0.32002752924982797


### TextRank

In [35]:
import math

import networkx as nx
import numpy as np


def build_textrank_matrix(tokenized_sentences):
    """
    Builds an N x N similarity matrix based on word overlap.

    tokenized_sentences: A list of lists containing cleaned, lemmatized words for each sentence.
    """
    N = len(tokenized_sentences)
    similarity_matrix = np.zeros((N, N))
    counter=0
    for i in range(N):
        # Start j from i + 1 to only calculate the upper triangle
        for j in range(i+1, N):
            # A sentence shouldn't vote for itself
            if i == j:
                continue
            counter+=1
            # Convert to sets to get unique words
            words_i = set(tokenized_sentences[i])
            words_j = set(tokenized_sentences[j])

            # |S_i ∩ S_j| - how many words overlapped in the 2 sentences
            overlap = len(words_i.intersection(words_j))

            if overlap == 0:
                similarity_matrix[i][j] = 0
                continue

            # log(|S_i|) + log(|S_j|)
            log_len_i = math.log(len(words_i)) if len(words_i) > 1 else 0
            log_len_j = math.log(len(words_j)) if len(words_j) > 1 else 0

            denominator = log_len_i + log_len_j

            # Prevent zero division if both sentences have 1 or 0 words
            # AND # Mirror the score to both sides of the diagonal
            if denominator == 0:
                similarity_matrix[i][j] = 0
                similarity_matrix[j][i] = 0
            else:
                score = overlap / denominator
                similarity_matrix[i][j] = score
                similarity_matrix[j][i] = score

    return similarity_matrix

In [36]:
def text_rank_scores(tokenized_sentences):
    # Build the N x N matrix and rank every pair of different sentences
    matrix = build_textrank_matrix(tokenized_sentences)

    # Convert the numpy matrix into a NetworkX graph
    similarity_graph = nx.from_numpy_array(matrix)

    # Run PageRank to get the final scores
    # This returns a dictionary formatted exactly like our TF scores: {0: 0.12, 1: 0.08, 2: 0.15...}
    textrank_scores = nx.pagerank(similarity_graph)

    # Sort the scores like before
    return sorted(textrank_scores.items(), key=lambda item: item[1], reverse=True)

#### nltk

In [37]:
# tokenize into words but keep the sentences separated
# list[ [sent_1_word_1, sent_1_word_2...], [sent_2_word_1, sent_2_word_2...] ]
nltk_tokenized_sentences = [nltk_tokenize(sent) for sent in nltk_raw_sentences]

nltk_textrank_scores_sorted = text_rank_scores(nltk_tokenized_sentences)
nltk_textrank_scores_sorted

[(2, 0.09495891624052138),
 (4, 0.09464923120212325),
 (14, 0.09117535874737943),
 (0, 0.09087808683926586),
 (9, 0.08386750252419274),
 (12, 0.0710331438078602),
 (6, 0.0689905619415409),
 (16, 0.06233873029897234),
 (7, 0.060393823341283494),
 (15, 0.05257438447159119),
 (11, 0.04746187349159178),
 (10, 0.044037208869793336),
 (8, 0.03963082717468058),
 (3, 0.0388146322229473),
 (5, 0.022250069437022298),
 (13, 0.019874370859356477),
 (1, 0.017071278529877377)]

In [38]:
# Generate summary using the exact same function we wrote earlier
nltk_textrank_summary = compose_summary_tf(nltk_raw_sentences, nltk_textrank_scores_sorted, SUMMARIZATION_COEF)
print(nltk_textrank_summary)

The Orbiter Discovery, OV-103, is considered eligible for listing in the National Register of Historic Places (NRHP) in the context of the U.S. Space Shuttle Program (1969-2011) under Criterion A in the areas of Space Exploration and Transportation and under Criterion C in the area of Engineering. Under Criterion A, Discovery is significant as the oldest of the three extant orbiter vehicles constructed for the Space Shuttle Program (SSP), the longest running American space program to date; she was the third of five orbiters built by NASA. Including her maiden voyage (launched August 30, 1984), Discovery flew to space thirty-nine times, more than any of the other four orbiters; she was also the first orbiter to fly twenty missions. In addition, Discovery was vital to the construction of the International Space Station (ISS); she flew thirteen of the thirty-seven total missions flown to the station by a U.S. Space Shuttle. The Space Shuttle orbiter also featured the first reusable TPS; a

In [39]:
print(f"Compression rate: {compression_rate(TEXT, nltk_textrank_summary)}")

Compression rate: 0.36889194769442535


#### SpaCy

In [52]:
spacy_tokenized_sentences = [spacy_tokenize(sent) for sent in spacy_raw_sentences]
spacy_textrank_scores_sorted = text_rank_scores(spacy_tokenized_sentences)
spacy_textrank_scores_sorted

[(2, 0.09925180847698599),
 (9, 0.09558500610059882),
 (0, 0.09469345699253187),
 (4, 0.09252995082159515),
 (12, 0.08320313119732378),
 (14, 0.08085919511154638),
 (16, 0.0685686817626695),
 (7, 0.0681475649461319),
 (6, 0.06209097701156618),
 (11, 0.045821778880465604),
 (15, 0.04283976537026399),
 (3, 0.04231420477981301),
 (10, 0.033562117087088776),
 (8, 0.030251734828256084),
 (13, 0.026638984162678965),
 (5, 0.024353716771954415),
 (1, 0.009287925698529414)]

In [53]:
spacy_textrank_summary = compose_summary_tf(spacy_raw_sentences, spacy_textrank_scores_sorted, SUMMARIZATION_COEF)
print(spacy_textrank_summary)

The Orbiter Discovery, OV-103, is considered eligible for listing in the National Register of Historic Places (NRHP) in the context of the U.S. Space Shuttle Program (1969-2011) under Criterion A in the areas of Space Exploration and Transportation and under Criterion C in the area of Engineering. Under Criterion A, Discovery is significant as the oldest of the three extant orbiter vehicles constructed for the Space Shuttle Program (SSP), the longest running American space program to date; she was the third of five orbiters built by NASA. Including her maiden voyage (launched August 30, 1984), Discovery flew to space thirty-nine times, more than any of the other four orbiters; she was also the first orbiter to fly twenty missions. In addition, Discovery was vital to the construction of the International Space Station (ISS); she flew thirteen of the thirty-seven total missions flown to the station by a U.S. Space Shuttle. According to Wayne Hale, a flight director from Johnson Space Cen

In [54]:
print(f"Compression rate: {compression_rate(TEXT, spacy_textrank_summary)}")

Compression rate: 0.40192704748795594


### LexRank

In [55]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity


def build_lexrank_scores(tokenized_sentences):
    # 1. sklearn expects strings, so we join our cleaned tokens back into sentences
    # e.g., ['orbiter', 'discovery', 'fly'] -> "orbiter discovery fly"
    cleaned_sentences = [" ".join(tokens) for tokens in tokenized_sentences]

    # 2. Automatically calculate TF-IDF for every word in every sentence
    vectorizer = TfidfVectorizer()
    tfidf_matrix = vectorizer.fit_transform(cleaned_sentences)

    # 3. Build the N x N similarity matrix using Cosine Similarity
    # (instead of the entire manual overlap loop)
    lexrank_matrix = cosine_similarity(tfidf_matrix)
    np.fill_diagonal(lexrank_matrix, 0.0) # avoid letting every sentence vote for itself

    # 4. Run the same PageRank engine
    lexrank_graph = nx.from_numpy_array(lexrank_matrix)
    lexrank_scores = nx.pagerank(lexrank_graph)

    return lexrank_scores

#### nltk

In [56]:
# Get scores and sort them
nltk_lexrank_scores_dict = build_lexrank_scores(nltk_tokenize(sent) for sent in nltk_raw_sentences)
nltk_lexrank_scores_sorted = sorted(nltk_lexrank_scores_dict.items(), key=lambda item: item[1], reverse=True)
nltk_lexrank_scores_sorted

[(4, 0.09331116041586933),
 (2, 0.08778685053557943),
 (6, 0.08351252149384435),
 (9, 0.08308306793369383),
 (14, 0.07777019898916913),
 (0, 0.06856236364877044),
 (15, 0.061115883020226726),
 (12, 0.05894851930271868),
 (10, 0.05861468485335352),
 (7, 0.057034579035389193),
 (16, 0.05655471970364001),
 (11, 0.05358452150837017),
 (8, 0.052897571414022626),
 (3, 0.036591525802631135),
 (5, 0.026697084408394717),
 (13, 0.023757127770943024),
 (1, 0.020177620163383675)]

In [57]:
# Generate the summary using your existing function
nltk_lexrank_summary = compose_summary_tf(nltk_raw_sentences, nltk_lexrank_scores_sorted, SUMMARIZATION_COEF)
print(nltk_lexrank_summary)

Under Criterion A, Discovery is significant as the oldest of the three extant orbiter vehicles constructed for the Space Shuttle Program (SSP), the longest running American space program to date; she was the third of five orbiters built by NASA. Including her maiden voyage (launched August 30, 1984), Discovery flew to space thirty-nine times, more than any of the other four orbiters; she was also the first orbiter to fly twenty missions. Discovery was the first shuttle to fly with the redesigned SRBs, a result of the Challenger accident, and the first shuttle to fly with the Phase II and Block I SSME. In addition, Discovery was vital to the construction of the International Space Station (ISS); she flew thirteen of the thirty-seven total missions flown to the station by a U.S. Space Shuttle. The Space Shuttle orbiter also featured the first reusable TPS; all previous spaceflight vehicles had a single-use, ablative heat shield.


In [58]:
print(f"Compression rate: {compression_rate(TEXT, nltk_lexrank_summary)}")

Compression rate: 0.3234686854783207


#### SpaCy

In [59]:
# Get scores and sort them
spacy_lexrank_scores_dict = build_lexrank_scores(spacy_tokenize(sent) for sent in spacy_raw_sentences)
spacy_lexrank_scores_sorted = sorted(spacy_lexrank_scores_dict.items(), key=lambda item: item[1], reverse=True)
spacy_lexrank_scores_sorted

[(9, 0.10165402803466228),
 (2, 0.1004928097325894),
 (4, 0.10041257306608191),
 (0, 0.07464466142611452),
 (6, 0.07314871368094071),
 (12, 0.07261508988960085),
 (14, 0.06714123672812818),
 (7, 0.06511389421447311),
 (16, 0.059054611418032914),
 (11, 0.0550007201581593),
 (15, 0.043554682024665094),
 (3, 0.04178386851310515),
 (8, 0.03969737910370327),
 (10, 0.03737944949339839),
 (5, 0.03109271172208415),
 (13, 0.02792564509766146),
 (1, 0.009287925696599267)]

In [60]:
# Generate the summary using your existing function
spacy_lexrank_summary = compose_summary_tf(spacy_raw_sentences, spacy_lexrank_scores_sorted, SUMMARIZATION_COEF)
print(spacy_lexrank_summary)

The Orbiter Discovery, OV-103, is considered eligible for listing in the National Register of Historic Places (NRHP) in the context of the U.S. Space Shuttle Program (1969-2011) under Criterion A in the areas of Space Exploration and Transportation and under Criterion C in the area of Engineering. Under Criterion A, Discovery is significant as the oldest of the three extant orbiter vehicles constructed for the Space Shuttle Program (SSP), the longest running American space program to date; she was the third of five orbiters built by NASA. Including her maiden voyage (launched August 30, 1984), Discovery flew to space thirty-nine times, more than any of the other four orbiters; she was also the first orbiter to fly twenty missions. Discovery was the first shuttle to fly with the redesigned SRBs, a result of the Challenger accident, and the first shuttle to fly with the Phase II and Block I SSME. In addition, Discovery was vital to the construction of the International Space Station (ISS

In [61]:
print(f"Compression rate: {compression_rate(TEXT, spacy_lexrank_summary)}")

Compression rate: 0.3788713007570544


## Abstractive summarization

Since text is self-supervising, we never use transformers from scratch. They are extremely expensive to train. 
- If the domain is to distant from the pre-trained model's original corpus, we might want to continue training (domain-adaptive pretraining), still might take hours to days depending on the model. 
- (If the data is labeled) If the model is familiar with the data, we only fine-tune - train the head on our labels
- If the task matches already, we just import and use as is

> **The exception**: when our "text" isn't text at all: protein sequences, DNA, music, chess moves, sensor streams, a proprietary log format. A model pretrained on English knows nothing useful about amino acids, so from-scratch is genuinely the answer there — and those models are often small enough to be affordable.

Based on the task, the family is **encoder-decoder**.

| we need to                             | family              | 2026 picks                                   |
| -------------------------------------- | ------------------- | -------------------------------------------- |
| understand (classify, tag, retrieve)   | **encoder**         | DeBERTa-v3, ModernBERT, DistilBERT (fast)    |
| generate                               | **decoder**         | GPT/Claude/Llama/Qwen                        |
| transform A → B (translate, summarize) | **encoder-decoder** | T5/Flan-T5, BART, opus-mt, NLLB              |
| embed sentences                        | **bi-encoder**      | `all-MiniLM-L6-v2` (fast), `bge-large`, `e5` |
| rerank retrieved results               | **cross-encoder**   | `ms-marco-MiniLM` cross-encoders             |

### Model selection

The task is very common and BART (Bidirectional and Auto-Regressive Transformer) is the most common pick. It is best for out-of-the-box abstractive summarization. It was trained on BookCorpus, English Wikipedia, CC-News, OpenWebText and Stories. Our data is written in the style of news and I don't see why space wouldn't be the topic it knows. Should be fine to use as is. If it turns out not enough, we might consider google's PEGASUS.

### BART

The next question is which model of BART to use specifically. There are multiple varieties. The source I considered was [hugging face](https://huggingface.co/models?pipeline_tag=summarization&sort=downloads). The name of the model already speaks for itself:

$$\underbrace{\text{facebook}}_{\text{who published it}} / \underbrace{\text{bart}}_{\text{architecture}} \underbrace{\text{-large-}}_{\text{size}} \underbrace{\text{cnn}}_{\text{fine-tuning dataset}}$$

At this stage, additional research with AI is needed to get to know the datasets better.

| Model | Trained On | Produces |
| :--- | :--- | :--- |
| facebook/bart-large-cnn | CNN/DailyMail | ~3 sentences, fairly extractive, faithful |
| google/pegasus-xsum | XSum (BBC) | highly abstractive, can make one-sentence or more |
| google/pegasus-cnn_dailymail | CNN/DailyMail | multi-sentence |
| sshleifer/distilbart-cnn-12-6 | CNN/DailyMail, distilled | same style, ~2× faster |

It is fair to say that a lot of models fit this task and the next decisive factor would then be **"which summary shape we want"**. Since we probably want a summary paragraph, `sshleifer/distilbart-cnn-12-6` looks like the one.

Now, the size. Looks like what's called "distilled" is similar to "mini/tiny" in CNNs. As usual, we start from the simplest, which I think more than enough for such small task. The model is chosen.

Now, the parameters:
- `max_length` / `min_length` - bounds on the output length in **tokens** (not words). Without these the model either stops early or hit the cap mid-word.
- `do_sample` - the determinism switch. `True` means sample from the probability distribution. Different output each run, more varied and creative, less faithful. Might hallucinate. `False` always take the highest-probability continuation. Same input → same output, every time, reliable.

For a summarization task, we would definitely want `do_sample=False`. The rest is left at defaults, exactly how the model was trained. It ensures its reliability and predictability. I see no reason to change anything until we encounter anything needed editing.

In [1]:
from transformers import AutoModelForSeq2SeqLM, AutoTokenizer

In [6]:
def count_tokens(tokenizer, text):
    """Token count for the current tokenizer."""
    return len(tokenizer.tokenize(text)) + tokenizer.num_special_tokens_to_add()

In [ ]:
DISTILBART_MODEL = "sshleifer/distilbart-cnn-12-6"                                     # ~1.2 GB, 306M params
distilbart_tokenizer = AutoTokenizer.from_pretrained(DISTILBART_MODEL)
distilbart_model = AutoModelForSeq2SeqLM.from_pretrained(DISTILBART_MODEL, use_safetensors=True)

# ---- check if the input size fits ----
distilbart_max_len = distilbart_model.config.max_position_embeddings                              # 1024 is this model's architectural limit
print(f"Max input token length for the architecture: {distilbart_max_len}")
print(f"Input token length of the input text: {count_tokens(distilbart_tokenizer, TEXT)}")

Max input token length for the architecture: 1024
Input token length of the input text: 595


The input length is less than the architecture holds, just proceeding.

In [ ]:
# ---- encode ----
distilbart_inputs = distilbart_tokenizer(TEXT, max_length=distilbart_max_len, truncation=True, return_tensors="pt")

# ---- generate ----
summary_ids = distilbart_model.generate(
    **distilbart_inputs,
    max_length=142,          # OUTPUT length cap, in tokens (model's own default)
    min_length=56,           # stops it emitting one thin sentence
    num_beams=4,             # beam search width
    length_penalty=2.0,      # >1 pushes toward longer output
    no_repeat_ngram_size=3,  # forbids repeating any 3-gram; BART repeats without it
    early_stopping=True,     # stop when all beams finish
    do_sample=False,         # deterministic: same input -> same summary
    forced_bos_token_id=0,   # controls which special token the decoder emits first (suppresses warning)
)

bart_summary = distilbart_tokenizer.decode(summary_ids[0], skip_special_tokens=True)
# strip the whitespaces + replace weird spacing before periods
bart_summary = (bart_summary) \
                            .strip() \
                            .replace(" .", ".")
print(bart_summary)
print(f"Compression rate: {compression_rate(TEXT, bart_summary)}")

OV-103 is considered eligible for listing in the National Register of Historic Places (NRHP) in the context of the U.S. Space Shuttle Program (1969-2011) under Criterion A in the areas of Space Exploration and Transportation. Because it has achieved significance within the past fifty years, Criteria Consideration G applies. Discovery flew to space thirty-nine times, more than any of the other four orbiters.
Compression rate: 0.14108740536820372


### PEGASUS

If the paragraph-style summary is not what we want, the alternative is a genuinely abstractive model. `PEGASUS` (Pre-training with Extracted Gap-sentences for abstractive summarization Sequence-to-sequence models, Google, 2020) was designed for short, genuinely abstractive summarization.

The pretraining data:
- C4 (Colossal Clean Crawled Corpus): a cleaned English version of Common Crawl, ~750 GB.
- HugeNews: 1.5 billion news articles (3.8 TB) collected 2013-2019.

But the more important difference is *how* it was pretrained. BART corrupts text in several ways (masking, deletion, shuffling) and learns to restore it - a general-purpose objective. PEGASUS uses **Gap Sentence Generation (GSG)**: it removes whole *sentences*  judged most important and trains the model to write them back. That is summarization practiced as the pretraining task itself, not an adaptation. It is why PEGASUS reaches good summaries with far less fine-tuning data.

Then the checkpoint. All PEGASUS checkpoints share the same architecture and differ in only three attributes: input limit, output length, and length penalty - i.e. they differ by the *shape of summary* they produce.

- `pegasus-xsum` -> XSum (BBC): every reference summary is a single sentence. Output is 
  one heavily-rewritten sentence.
- `pegasus-cnn_dailymail` -> multi-sentence, closer to BART's behavior.
- `pegasus-large` -> the un-fine-tuned base, meant for fine-tuning on your own data.

Since the point here is contrast with the extractive algorithms and near-extractive BART output, `pegasus-cnn_dailymail` is the right pick.

The parameters, all inherited from the CNN_DM's checkpoint's own configuration.

In [ ]:
from transformers import AutoModelForSeq2SeqLM, AutoTokenizer

In [ ]:
# !pip install sentencepiece

In [ ]:
PEGASUS_CNN_DM_MODEL = "google/pegasus-cnn_dailymail"                 # ~2.2 GB, 568M params
pegasus_cnn_dm_tokenizer = AutoTokenizer.from_pretrained(PEGASUS_CNN_DM_MODEL)
pegasus_cnn_dm_model = AutoModelForSeq2SeqLM.from_pretrained(PEGASUS_CNN_DM_MODEL, use_safetensors=True)

# ---- check if the input size fits ----
pegasus_cnn_dm_max_len = pegasus_cnn_dm_model.config.max_position_embeddings                              # 1024 is this model's architectural limit
print(f"Max input token length for the architecture: {pegasus_cnn_dm_max_len}")
print(f"Input token length of the input text: {count_tokens(pegasus_cnn_dm_tokenizer, TEXT)}")

Some weights of PegasusForConditionalGeneration were not initialized from the model checkpoint at google/pegasus-cnn_dailymail and are newly initialized: ['model.decoder.embed_positions.weight', 'model.encoder.embed_positions.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Max input token length for the architecture: 1024
Input token length of the input text: 567


> **Note**: what if input tokens were greater than the model's capability (> `pegasus_cnn_dm_max_len`)?
> If the inputs the model expects are usually within the model's input capacity range, we might just go away with truncating treating the current one as an outlier. But since we have multiple extractive summarization models built, we'd use one of them to extract the most important information. We would sort them by importance and kept adding sentences until the model's limit has been exceeded.

In [ ]:
# ---- encode ----
pegasus_cnn_dm_inputs = pegasus_cnn_dm_tokenizer(TEXT, max_length=pegasus_cnn_dm_max_len, truncation=True, return_tensors="pt")

# ---- generate: CNN_DM's own tuned settings ----
summary_ids_x = pegasus_cnn_dm_model.generate(
    **pegasus_cnn_dm_inputs,
    max_length=64,           # CNN_DM target is ONE sentence, not a paragraph
    num_beams=8,             # CNN_DM was tuned with 8 beams, not 4
    length_penalty=0.6,      # <1 pushes SHORTER (BART's 2.0 pushed longer)
    early_stopping=True,
    do_sample=False,
    # no min_length      - forcing 56 tokens would destroy the one-sentence style
)

pegasus_summary = pegasus_cnn_dm_tokenizer.decode(summary_ids_x[0], skip_special_tokens=True)
pegasus_summary = (pegasus_summary) \
                                .strip() \
                                .replace(" .", ".") \
                                .replace("<n>", "\n")

# remove new line markers before calculating the compression rate
clean_summary = pegasus_summary.replace('\n', '')
print(f"Compression rate: {compression_rate(TEXT, clean_summary)}")
print(pegasus_summary)

Compression rate: 0.10839642119752237
The Orbiter Discovery, OV-103, is considered eligible for listing in the National Register of Historic Places.
Under Criterion A, Discovery is significant as the oldest of the three extant orbiter vehicles constructed for the Space Shuttle Program.
Under Criterion C, Discovery is significant as a feat of engineering


## Conclusion

### Pre-processing: nltk vs SpaCy

1. `nltk`'s tokenizer follows Penn Treebank rules from the 1990s, where hyphenated compounds stay whole, `SpaCy` follows a more modern approach. For the hyphenated expressions, `SpaCy` is right, for the years - I stand with `nltk`:

|`nltk`|`spacy`|
|------|-------|
|"1969-2011"|"1969", "2011"|
|"defense-related"|"defense", "related"|
| "cargo-carrying"|"cargo", "carrying"|

2. `SpaCy` has a broader range of punctuation without having to add any manually.

3. `SpaCy` didn't include some of the word numbers: "fifty", "three", "four", "third", "thirty", "thirty-nine", "five" etc., but it did include "thirteen" for no reason and "thirty" and "seven" because they were hyphenated: "thirty-seven". `nltk` included all of them as separate tokens. I would also stand with `nltk` here.

4. `nltk` is behind the POS recognition, if a word can appear as both verb and noun, it thinks it might be a noun because the previous words were nouns. It's simply less sophisticated, it's not a deep NN after all:

|`nltk`|`spacy`|
|------|-------|
|"applies"|"apply"|
|"redesigned"|"redesign"|

5. `nltk` doesn't feature stop words: "also", "last", "well", "although", whereas `SpaCy` does. We could exclude `wordnet.ADV` to omit the adverbs in `nltk` as well, but that's arguable.

I would use SpaCy. Easier, more relevant and precise, though some custom tweaks would be necessary regardless.

### Summary comparison

1. `nltk -> lead-3`:
'The Orbiter Discovery, OV-103, is considered eligible for listing in the National Register of Historic Places (NRHP) in the context of the U.S. Space Shuttle Program (1969-2011) under Criterion A in the areas of Space Exploration and Transportation and under Criterion C in the area of Engineering. Because it has achieved significance within the past fifty years, Criteria Consideration G applies. Under Criterion A, Discovery is significant as the oldest of the three extant orbiter vehicles constructed for the Space Shuttle Program (SSP), the longest running American space program to date; she was the third of five orbiters built by NASA.'

2. `SpaCy -> lead-3`:
Identical to #1.

3. `nltk -> TF`:
'Including her maiden voyage (launched August 30, 1984), Discovery flew to space thirty-nine times, more than any of the other four orbiters; she was also the first orbiter to fly twenty missions. Discovery was the first shuttle to fly with the redesigned SRBs, a result of the Challenger accident, and the first shuttle to fly with the Phase II and Block I SSME. In addition, Discovery was vital to the construction of the International Space Station (ISS); she flew thirteen of the thirty-seven total missions flown to the station by a U.S. Space Shuttle. She was the first orbiter to dock to the ISS, and the first to perform an exchange of a resident crew. The Space Shuttle orbiter also featured the first reusable TPS; all previous spaceflight vehicles had a single-use, ablative heat shield.'

4. `SpaCy -> TF`:
'Under Criterion A, Discovery is significant as the oldest of the three extant orbiter vehicles constructed for the Space Shuttle Program (SSP), the longest running American space program to date; she was the third of five orbiters built by NASA. Including her maiden voyage (launched August 30, 1984), Discovery flew to space thirty-nine times, more than any of the other four orbiters; she was also the first orbiter to fly twenty missions. Discovery was the first shuttle to fly with the redesigned SRBs, a result of the Challenger accident, and the first shuttle to fly with the Phase II and Block I SSME. Discovery also carried the Hubble Space Telescope to orbit and performed two of the five servicing missions to the observatory. In addition, Discovery was vital to the construction of the International Space Station (ISS); she flew thirteen of the thirty-seven total missions flown to the station by a U.S. Space Shuttle.'

5. `nltk -> TextRank -> TF`:
'The Orbiter Discovery, OV-103, is considered eligible for listing in the National Register of Historic Places (NRHP) in the context of the U.S. Space Shuttle Program (1969-2011) under Criterion A in the areas of Space Exploration and Transportation and under Criterion C in the area of Engineering. Under Criterion A, Discovery is significant as the oldest of the three extant orbiter vehicles constructed for the Space Shuttle Program (SSP), the longest running American space program to date; she was the third of five orbiters built by NASA. Including her maiden voyage (launched August 30, 1984), Discovery flew to space thirty-nine times, more than any of the other four orbiters; she was also the first orbiter to fly twenty missions. In addition, Discovery was vital to the construction of the International Space Station (ISS); she flew thirteen of the thirty-seven total missions flown to the station by a U.S. Space Shuttle. The Space Shuttle orbiter also featured the first reusable TPS; all previous spaceflight vehicles had a single-use, ablative heat shield.'

6. `SpaCy -> TextRank -> TF`:
'The Orbiter Discovery, OV-103, is considered eligible for listing in the National Register of Historic Places (NRHP) in the context of the U.S. Space Shuttle Program (1969-2011) under Criterion A in the areas of Space Exploration and Transportation and under Criterion C in the area of Engineering. Under Criterion A, Discovery is significant as the oldest of the three extant orbiter vehicles constructed for the Space Shuttle Program (SSP), the longest running American space program to date; she was the third of five orbiters built by NASA. Including her maiden voyage (launched August 30, 1984), Discovery flew to space thirty-nine times, more than any of the other four orbiters; she was also the first orbiter to fly twenty missions. In addition, Discovery was vital to the construction of the International Space Station (ISS); she flew thirteen of the thirty-seven total missions flown to the station by a U.S. Space Shuttle. According to Wayne Hale, a flight director from Johnson Space Center, the Space Shuttle orbiter represents a “huge technological leap from expendable rockets and capsules to a reusable, winged, hypersonic, cargo-carrying spacecraft.”'

7. `nltk -> LexRank -> TF`:
'Under Criterion A, Discovery is significant as the oldest of the three extant orbiter vehicles constructed for the Space Shuttle Program (SSP), the longest running American space program to date; she was the third of five orbiters built by NASA. Including her maiden voyage (launched August 30, 1984), Discovery flew to space thirty-nine times, more than any of the other four orbiters; she was also the first orbiter to fly twenty missions. Discovery was the first shuttle to fly with the redesigned SRBs, a result of the Challenger accident, and the first shuttle to fly with the Phase II and Block I SSME. In addition, Discovery was vital to the construction of the International Space Station (ISS); she flew thirteen of the thirty-seven total missions flown to the station by a U.S. Space Shuttle. The Space Shuttle orbiter also featured the first reusable TPS; all previous spaceflight vehicles had a single-use, ablative heat shield.'


8. `SpaCy -> LexRank -> TF`:
The Orbiter Discovery, OV-103, is considered eligible for listing in the National Register of Historic Places (NRHP) in the context of the U.S. Space Shuttle Program (1969-2011) under Criterion A in the areas of Space Exploration and Transportation and under Criterion C in the area of Engineering. Under Criterion A, Discovery is significant as the oldest of the three extant orbiter vehicles constructed for the Space Shuttle Program (SSP), the longest running American space program to date; she was the third of five orbiters built by NASA. Including her maiden voyage (launched August 30, 1984), Discovery flew to space thirty-nine times, more than any of the other four orbiters; she was also the first orbiter to fly twenty missions. Discovery was the first shuttle to fly with the redesigned SRBs, a result of the Challenger accident, and the first shuttle to fly with the Phase II and Block I SSME. In addition, Discovery was vital to the construction of the International Space Station (ISS); she flew thirteen of the thirty-seven total missions flown to the station by a U.S. Space Shuttle.


9. `BART` (distilbart-cnn-12-6):
'OV-103 is considered eligible for listing in the National Register of Historic Places (NRHP) in the context of the U.S. Space Shuttle Program (1969-2011) under Criterion A in the areas of Space Exploration and Transportation. Because it has achieved significance within the past fifty years, Criteria Consideration G applies. Discovery flew to space thirty-nine times, more than any of the other four orbiters.'

10. `PEGASUS` (pegasus-cnn_dailymail):
'The Orbiter Discovery, OV-103, is considered eligible for listing in the National Register of Historic Places.
Under Criterion A, Discovery is significant as the oldest of the three extant orbiter vehicles constructed for the Space Shuttle Program.
Under Criterion C, Discovery is significant as a feat of engineering'

|Model/algorithm|Pre-processor|Compression rate (ratio of original text preserved)|Summarization|
|------------------|-------------|----------------|-------------------------|
|lead-3 (baseline) | nltk        |0.2216          |extractive|
|lead-3 (baseline) | SpaCy       |0.2216          |extractive|
|TF                 | nltk        |0.2743         |extractive|
|TF                 | SpaCy       |0.3200          |extractive|
|TextRank           | nltk        |0.3689          |extractive|
|TextRank           | SpaCy       |0.4019          |extractive|
|LexRank           | nltk        |0.3235          |extractive|
|LexRank           | SpaCy       |0.3789          |extractive|
|BART           | distilbart-cnn-12-6       |0.1411          |abstractive|
|PEGASUS           | pegasus-cnn_dailymail       |0.1084          |abstractive|

#### Extractive summarization

The given text itself already looks a bit like a summary. Might be because either because it is or just a marketing strategy to get the user involved into the topic immediately. For this very reason, the `lead-3` baseline does well, but I wouldn't say it's the best. It's arguably worse than TF, but wins in compression rate within extractive summarization. 

With each next model summary started looking better. For example, direct usage of TF included criterions right away. It felt abrupt and confusing. Subjectively, I'd say LexRank, being the most complex algorithm within the extractive approach, has surpassed the rest. The `nltk` and `SpaCy` versions are identical.

#### Abstractive summarization

The BART's summary delivers the same thought far more concisely. It is 2.67 times shorter than the established extractive model.

The PEGASUS's summary captured the main part perfectly, also presenting it in a visually pleasant manner. Additionally, the compression is the lowest (almost 3.5 more compressed than the best extractive).


#### Final choice

- for extractive: SpaCy -> LexRank -> TF
- for abstractive: PEGASUS
- for abstractive (larger texts): SpaCy -> LexRank -> TF -> PEGASUS